### PR TITLE
Secure Tunnel V2 WebSocket Protocol support

### PR DESCRIPTION
**Pull request by sbSteveK**
- Refactor Secure Tunnel to use new Secure Tunnel API.
- 
- V2WebSocketProtocol support for Secure Tunnel.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: secure_tunneling/source/SecureTunnel.cpp [hunk 1]
```diff
@@ -10,6 +10,189 @@ namespace Aws
...
+        }
+
+        //***********************************************************************************************************************
+        /*                                              Message */
+        //***********************************************************************************************************************
+
+        Message::Message(const aws_secure_tunnel_message_view &message, Crt::Allocator *allocator) noexcept
+            : m_allocator(allocator)
+        {
+            AWS_ZERO_STRUCT(m_payload);
```
Comment (xiazhvera) [inline secure_tunneling/source/SecureTunnel.cpp:54]: I'm not sure if you can do a AWS_ZERO_STRUCT to a Crt::Optional. Probably need a further testing on it...

Comment (sbSteveK) [reply secure_tunneling/source/SecureTunnel.cpp]: Cool! Remove the Zero Struct for both optional members.

**Review by xiazhvera — APPROVED** [review]
Some trivial suggestions. 
The only concern I have was about how was the callback get handled during destruction. Other than that, the PR looks good to me.

File: secure_tunneling/source/SecureTunnel.cpp [hunk 2]
```diff
@@ -322,33 +607,104 @@ namespace Aws
...
+                    secureTunnel->m_OnConnectionFailure(secureTunnel, error_code);
+                }
             }
         }
 
-        void SecureTunnel::s_OnConnectionShutdown(void *user_data)
+        void SecureTunnel::s_OnConnectionShutdown(int error_code, void *user_data)
         {
-            auto *secureTunnel = static_cast<SecureTunnel *>(user_data);
+            SecureTunnel *secureTunnel = reinterpret_cast<SecureTunnel *>(user_data);
```
[the remark this answers is not in the document]
Comment (sbSteveK) [reply secure_tunneling/source/SecureTunnel.cpp]: Nope, changed to static_cast!

File: documents/Secure_Tunnel_Userguide.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+// Messages received on a stream will now be printed to stdout.
+```
+
+# How to Start and Stop
+
+## Start
+Invoking Start() on the Secure Tunnel Client will put it into an active state where it recurrently establishes a connection to the configured Secure Tunnel endpoint using the provided [Client Access Token](https://docs.aws.amazon.com/iot/latest/developerguide/secure-tunneling-concepts.html). If a [Client Token](https://docs.aws.amazon.com/iot/latest/developerguide/secure-tunneling-concepts.html) is provided, the Secure Tunnel Client will use it but in the event one is not provided, the Secure Tunnel Client will automatically generate one for use on a reconnection attempts. The Client Token for any initial connection to the Secure Tunnel service MUST be unique. Reusing a Client Token from a previous connection will result in a failed connection to the Secure Tunnel Service.
+```
```
Comment (TwistedTwigleg) [inline documents/Secure_Tunnel_Userguide.md:113]: Minor: Add `cpp` to code block

File: documents/Secure_Tunnel_Userguide.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+// Start the secure tunnel connection
+if (!secureTunnel->Start())
+{
+    fprintf("Failed to start Secure Tunnel\n");
+    return -1;
+}
+```
+
+## Stop
+Invoking Stop() on the Secure Tunnel Client breaks the current connection (if any) and moves the client into an idle state.
```
Comment (TwistedTwigleg) [inline documents/Secure_Tunnel_Userguide.md:139]: I might replace `(if any)` to simply `if one exists`.

Minor: Put `Stop()` in a code block, add `cpp` to the big code block below.

File: documents/Secure_Tunnel_Userguide.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+```
+if(!secureTunnel->Stop()){
+    fprintf(stdout, "Failed to stop the Secure Tunnel connection session. Exiting..\n");
+}
+```
+
+# Secure Tunnel Operations
+
+## Send Message
+The SendMessage operation takes a description of the Message you wish to send and returns a success/failure in the synchronous logic that kicks off the Send Message operation. When the message is fully written to the socket, the OnSendDataComplete callback will be invoked.
```
Comment (TwistedTwigleg) [inline documents/Secure_Tunnel_Userguide.md:149]: Minor: SendMessage -> `SendMessage()` and `Send Message` -> `SendMessage()`. Ditto with `OnSendDataComplete`.

Add `cpp` to the code block below.

**Review by TwistedTwigleg — APPROVED** [review]
I only looked through the documentation and sample README, but it looks good. I just had a few nitpicks but they are extremely minor.

**Pull request merged**